server/data_common/rwlock.py

```diff
@@ -88,6 +90,12 @@ def w_acquire(self):
     def w_release(self):
         self.w_lock.release()
 
+    def w_demote(self):
```

Comment:
this does not release the write lock, and the caller in matrix_loader does not either.   That will work OK if it is called from within a `with lock.w_locked() as lock:...`, but won't work as currently called.

Reply:
The point is not to release the w_lock.   w_lock is perhaps not a great name here, since the reader also hold the w_lock.   The effect I'm going for here is to get into a state that would occur if I had called this atomically.
   rw_lock.w_release()
   rw.local.r_acquire()

It's a little frustrating that python standard library does not have a rwlock.  They are tricky to get right.  I just realized that this implementation has a deadlock possibility.  So, I need to find a better way.   However demoting the writer to a reader is the right solution I believe

Reply:
you are absolutely right!   apologies, will resolve this conversation.